File: crates/kernel-consumer/tests/consumer.rs
//! Runtime evidence that an ordinary Rust crate can consume the packaged kernel.

use agent_ix_semantic_kernel::{
	ClauseRef, ConstraintDecl, FieldDecl, Identifier, KernelScalar, Multiplicity,
	MultiplicityLower, MultiplicityOrdered, MultiplicityUnique, MultiplicityUpper,
	OperationDecl, RelationDecl, TypeRef, TypeRefTarget,
};

const MULTIPLICITY_GOLDEN: &str = include_str!("../fixtures/PAR-0001.json");
const CONSTRAINT_GOLDEN: &str = include_str!("../fixtures/PAR-0027.json");

fn constructed_field() -> FieldDecl {
	let multiplicity = Multiplicity::try_new(
		MultiplicityLower::try_new(1).expect("lower is valid"),
		Some(MultiplicityUpper::try_new(1).expect("upper is valid")),
		MultiplicityOrdered::try_new(false).expect("ordered is valid"),
		MultiplicityUnique::try_new(false).expect("unique is valid"),
	)
	.expect("multiplicity constructs");
	let target = TypeRef::try_new(
		TypeRefTarget::KernelScalar(KernelScalar::String),
		Some(multiplicity),
		None,
		None,
	)
	.expect("type reference constructs");
	let constraint: ConstraintDecl = serde_json::from_str(r#"{"keyword":"min","value":0}"#)
		.expect("the primitive union constraint constructs");
	FieldDecl::try_new(
		Identifier::try_new("unitSymbol".to_owned()).expect("identifier constructs"),
		target,
		None,
		None,
		None,
		None,
		Some(vec![constraint]),
	)
	.expect("field declaration constructs")
}

macro_rules! rejects {
	($type:ty, $json:expr $(,)?) => {
		serde_json::from_str::<$type>($json)
			.expect_err("the forbidden document must be rejected")
			.to_string()
	};
}

/// TC-1544: constructs through the generated crate's public root and pins the
/// imports FR-089 requires a consumer to use.
#[test]
fn tc_1544_constructs_a_field_and_imports_the_kernel_surface() {
	let field = constructed_field();
	assert_eq!(field.name.get(), "unitSymbol");
	assert!(field.r#type.multiplicity.is_some());
	let _: Option<(RelationDecl, OperationDecl, ClauseRef, ConstraintDecl)> = None;
}

/// TC-1545: the consumer deserializes a positive FR-090 corpus document.
#[test]
fn tc_1545_deserializes_the_shared_positive_golden() {
	let golden: serde_json::Value = serde_json::from_str(MULTIPLICITY_GOLDEN)
		.expect("PAR-0001 is JSON");
	let value: Multiplicity = serde_json::from_value(
		golden.get("instance").expect("PAR-0001 carries an instance").clone(),
	)
		.expect("PAR-0001 deserializes through the packaged crate");
	assert_eq!(*value.lower.get(), 0);
	assert_eq!(*value.upper.expect("golden carries upper").get(), 1);

	let golden: serde_json::Value = serde_json::from_str(CONSTRAINT_GOLDEN)
		.expect("PAR-0027 is JSON");
	let value: ConstraintDecl = serde_json::from_value(
		golden.get("instance").expect("PAR-0027 carries an instance").clone(),
	)
	.expect("PAR-0027 preserves the numeric constraint operand");
	assert_eq!(
		serde_json::to_value(value)
			.expect("constraint serializes")["value"]
			.as_f64(),
		Some(0.0),
	);
}

/// TC-1546: every closed grammar class named by FR-089 is refused with its
/// generated error, rather than merely producing a generic failure.
#[test]
fn tc_1546_rejects_each_closed_grammar_class() {
	let unknown = rejects!(FieldDecl,
		r#"{"name":"unitSymbol","type":{"target":"String"},"undeclared":1}"#,
	);
	assert!(unknown.contains("unknown field `undeclared`"), "{unknown}");

	let missing = rejects!(FieldDecl, r#"{"name":"unitSymbol"}"#);
	assert!(missing.contains("missing field `type`"), "{missing}");

	let scalar = rejects!(KernelScalar, r#""NotAKernelScalar""#);
	assert!(scalar.contains("unknown variant `NotAKernelScalar`"), "{scalar}");

	let constraint = rejects!(ConstraintDecl, r#"{"keyword":"notAKeyword"}"#);
	assert!(constraint.contains("data did not match any variant"), "{constraint}");

	let minimum = rejects!(
		Multiplicity,
		r#"{"lower":-1,"ordered":false,"unique":false}"#,
	);
	assert!(minimum.contains("keyword min"), "{minimum}");
}

/// TC-1547: identity and source provenance are read from the package itself.
#[test]
fn tc_1547_reads_identity_and_provenance_from_the_package() {
	assert_eq!(kernel_consumer::package_identity(), "agent-ix/semantic-kernel");
	assert_eq!(agent_ix_semantic_kernel::provenance::SOURCE_VERSION, "0.2.0");
	assert_eq!(
		agent_ix_semantic_kernel::provenance::SOURCE_DIGEST,
		"sha256:0000000000000000000000000000000000000000000000000000000000000000",
	);
}

File: packages/semantic-kernel/examples/typescript/consumer.ts
/** An ordinary, dependency-free TypeScript consumer of the kernel (FR-089). */

import {
	PROVENANCE,
	TYPE_IDENTITY,
	validateConstraintDecl,
	validateFieldDecl,
	validateKernelScalar,
	validateMultiplicity,
} from "@agent-ix/semantic-agent-ix__semantic-kernel";
import type {
	ClauseRef,
	ConstraintDecl,
	FieldDecl,
	KernelScalar,
	Multiplicity,
	OperationDecl,
	RelationDecl,
	TypeRef,
} from "@agent-ix/semantic-agent-ix__semantic-kernel";
import PAR_0001 from "./PAR-0001.json" with { type: "json" };
import PAR_0027 from "./PAR-0027.json" with { type: "json" };

let assertions = 0;

function assert(condition: unknown, message: string): asserts condition {
	assertions += 1;
	if (!condition) throw new Error(message);
}

function rejected(result: { readonly ok: boolean }, name: string): void {
	assert(
		!result.ok,
		`${name} was accepted where the kernel grammar rejects it`,
	);
}

/** Executes every FR-089 demonstration and returns its non-zero assertion count. */
export function run(): number {
	const field: FieldDecl = {
		name: "unitSymbol",
		type: {
			target: "String",
			multiplicity: { lower: 1, upper: 1, ordered: false, unique: false },
		},
		constraints: [{ keyword: "min", value: 0 }],
	};
	const constructed = validateFieldDecl(field);
	assert(constructed.ok, "a valid FieldDecl with multiplicity was rejected");

	// These declarations are deliberately named through the public type surface:
	// removing or renaming any required import makes this consumer fail to check.
	const imported: readonly [
		TypeRef | undefined,
		Multiplicity | undefined,
		RelationDecl | undefined,
		OperationDecl | undefined,
		ClauseRef | undefined,
		ConstraintDecl | undefined,
		KernelScalar | undefined,
	] = [
		undefined,
		undefined,
		undefined,
		undefined,
		undefined,
		undefined,
		undefined,
	];
	assert(imported.length === 7, "the required public types were not imported");

	const positive = validateMultiplicity(PAR_0001.instance);
	assert(positive.ok, "PAR-0001 did not deserialize");
	if (positive.ok) assert(positive.value.upper === 1, "PAR-0001 lost upper");
	const constraint = validateConstraintDecl(PAR_0027.instance);
	assert(constraint.ok, "PAR-0027 did not preserve its numeric operand");
	if (constraint.ok)
		assert(constraint.value.keyword === "min", "PAR-0027 lost its keyword");

	rejected(
		validateFieldDecl({ ...field, undeclared: 1 }),
		"an undeclared member on a reject record",
	);
	rejected(
		validateFieldDecl({ name: "unitSymbol" }),
		"a missing required member",
	);
	rejected(validateKernelScalar("NotAKernelScalar"), "an unknown KernelScalar");
	rejected(
		validateConstraintDecl({ keyword: "notAKeyword" }),
		"an unknown ConstraintDecl keyword",
	);
	rejected(validateMultiplicity({ lower: -1 }), "a minimum-bound violation");

	assert(
		TYPE_IDENTITY.FieldDecl === "ix://agent-ix/semantic-core/FieldDecl",
		"the package identity metadata is wrong",
	);
	assert(PROVENANCE.sourceVersion === "0.2.0", "the source version is wrong");
	assert(
		PROVENANCE.sourceDigest ===
			"sha256:0000000000000000000000000000000000000000000000000000000000000000",
		"the source digest is wrong",
	);
	assert(assertions > 0, "the consumer executed no assertions");
	return assertions;
}
